File: src-tauri/src/commands/media_library.rs
use crate::db::models::{MediaLibraryCategory, MediaLibraryCategoryInput, MediaLibraryItem, MediaLibraryItemInput};
use crate::error::AppError;
use crate::state::AppState;
use crate::utils::catcher::catcher;

#[tauri::command]
#[specta::specta]
pub fn get_media_library_categories(
    language: String,
    state: tauri::State<'_, AppState>,
) -> Result<Vec<MediaLibraryCategory>, AppError> {
    let (conn, err) = catcher(state.db.get());
    if let Some(e) = err {
        return Err(e);
    }
    let conn = conn.unwrap();
    crate::db::queries::media_library::get_categories(&conn, &language)
}

#[tauri::command]
#[specta::specta]
pub fn upsert_media_library_category(
    input: MediaLibraryCategoryInput,
    state: tauri::State<'_, AppState>,
) -> Result<i64, AppError> {
    let (conn, err) = catcher(state.db.get());
    if let Some(e) = err {
        return Err(e);
    }
    let conn = conn.unwrap();
    crate::db::queries::media_library::upsert_category(&conn, input)
}

#[tauri::command]
#[specta::specta]
pub fn delete_media_library_category(
    id: i64,
    state: tauri::State<'_, AppState>,
) -> Result<(), AppError> {
    let (conn, err) = catcher(state.db.get());
    if let Some(e) = err {
        return Err(e);
    }
    let conn = conn.unwrap();
    crate::db::queries::media_library::delete_category(&conn, id)
}

#[tauri::command]
#[specta::specta]
pub fn get_media_library_items(
    category_id: i64,
    state: tauri::State<'_, AppState>,
) -> Result<Vec<MediaLibraryItem>, AppError> {
    let (conn, err) = catcher(state.db.get());
    if let Some(e) = err {
        return Err(e);
    }
    let conn = conn.unwrap();
    crate::db::queries::media_library::get_items_by_category(&conn, category_id)
}

#[tauri::command]
#[specta::specta]
pub fn upsert_media_library_item(
    input: MediaLibraryItemInput,
    state: tauri::State<'_, AppState>,
) -> Result<i64, AppError> {
    let (conn, err) = catcher(state.db.get());
    if let Some(e) = err {
        return Err(e);
    }
    let conn = conn.unwrap();
    crate::db::queries::media_library::upsert_item(&conn, input)
}

#[tauri::command]
#[specta::specta]
pub fn delete_media_library_item(
    id: i64,
    state: tauri::State<'_, AppState>,
) -> Result<(), AppError> {
    let (conn, err) = catcher(state.db.get());
    if let Some(e) = err {
        return Err(e);
    }
    let conn = conn.unwrap();
    crate::db::queries::media_library::delete_item(&conn, id)
}

#[tauri::command]
#[specta::specta]
pub fn search_media_library_items(
    query: String,
    state: tauri::State<'_, AppState>,
) -> Result<Vec<MediaLibraryItem>, AppError> {
    let (conn, err) = catcher(state.db.get());
    if let Some(e) = err {
        return Err(e);
    }
    let conn = conn.unwrap();
    crate::db::queries::media_library::search_library_items(&conn, &query)
}

File: apps/desktop/src/lib/tauri/media-library.ts
import { invoke } from "@tauri-apps/api/core";
import type {
  MediaLibraryCategory,
  MediaLibraryCategoryInput,
  MediaLibraryItem,
  MediaLibraryItemInput,
} from "../bindings";

async function tauriInvoke<T>(
  command: string,
  args?: Record<string, unknown>,
): Promise<T> {
  return invoke<T>(command, args);
}

// Media Library
export async function getMediaLibraryCategories(language: string): Promise<MediaLibraryCategory[]> {
  return tauriInvoke<MediaLibraryCategory[]>("get_media_library_categories", { language });
}

export async function upsertMediaLibraryCategory(input: MediaLibraryCategoryInput): Promise<number> {
  return tauriInvoke<number>("upsert_media_library_category", { input });
}

export async function deleteMediaLibraryCategory(id: number): Promise<void> {
  return tauriInvoke<void>("delete_media_library_category", { id });
}

export async function getMediaLibraryItems(categoryId: number): Promise<MediaLibraryItem[]> {
  return tauriInvoke<MediaLibraryItem[]>("get_media_library_items", { categoryId });
}

export async function getMediaLibraryItemsByDate(
  categoryId: number,
  date: string | null,
): Promise<MediaLibraryItem[]> {
  return tauriInvoke<MediaLibraryItem[]>("get_media_library_items_by_date", { categoryId, date });
}

export async function getMediaLibraryItemDates(categoryId: number): Promise<string[]> {
  return tauriInvoke<string[]>("get_media_library_item_dates", { categoryId });
}

export async function upsertMediaLibraryItem(input: MediaLibraryItemInput): Promise<number> {
  return tauriInvoke<number>("upsert_media_library_item", { input });
}

export async function deleteMediaLibraryItem(id: number): Promise<void> {
  return tauriInvoke<void>("delete_media_library_item", { id });
}

export async function searchMediaLibraryItems(query: string): Promise<MediaLibraryItem[]> {
  return tauriInvoke<MediaLibraryItem[]>("search_media_library_items", { query });
}

export async function getScheduledMediaItem(categoryId: number, date: string): Promise<MediaLibraryItem | null> {
  return tauriInvoke<MediaLibraryItem | null>("get_scheduled_media_item", { categoryId, date });
}
